### DISCUSSION TITLE
Remove everything except images

### DISCUSSION THREAD

**Discussion by nsklei** [pymupdf/PyMuPDF#3425] [Looking for help] [answered]
I am looking for a solution to remove all elements from a PDF (text, paths ...) except images.

My current solution extracts all images from the document and inserts them inside a new document.
Is there any simpler way to solve this problem, because extracting and inserting the images feels like a workaround?

**Comment by JorjMcKie** [comment]
Sure there is a canonical way: Use redaction annotations!
1. add a redaction for the full page: `page.add_redact_annot(page.rect)`
2. apply redactions using an option that ignores images for the removal: `page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE)`.
3. Save the file with sufficient garbage collection and compression: `doc.ez_save("redacted.pdf")`.

Since the recent version, all 3 categories, images, graphics and text can be selectively kept or removed.

**Reply by nsklei** [reply]
I already tried that, but it did not work completely.
Most of the text elements were removed, but some of them and also the path elements are still present.

```
import fitz

doc = fitz.open("Testfile.pdf")

for page in doc:
    page.add_redact_annot(page.rect)
    page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE)

doc.ez_save("Testfile_redacted.pdf")
doc.close()
```

[Testfile.pdf](https://github.com/pymupdf/PyMuPDF/files/15167282/Testfile.pdf)

**Reply by JorjMcKie** [reply, marked as answer]
That happen when the PDF was sloppily created: with text or drawings not completely inside page.rect.
For drawings, you can / should use the option that erase even when only parts are overlapping.
I believe `fitz.PDF_REDACT_LINE_ART_REMOVE_IF_TOUCHED` (=2).
For stubborn text, simply increase the redaction rectangle like so `page.rect + (-20, -20, 20, 20)` which is a rect 20 points larger in every direction.

**Reply by nsklei** [reply]
Yep, that worked. Thank you :)
